src/probability/upcrossing.lean

```diff
@@ -0,0 +1,897 @@
+/-
+Copyright (c) 2022 Kexing Ying. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kexing Ying
+-/
+import probability.hitting_time
+import probability.martingale
+
+/-!
+
+# Doob's upcrossing estimate
+
+Given a discrete real-valued submartingale $(f_n)_{n \in \mathbb{N}}$, denoting $U_N(a, b)$ the
+number of times $f_n$ crossed from below $a$ to above $b$ before time $N$, Doob's upcrossing
+estimate (also known as Doob's inequality) states that
+$$(b - a) \mathbb{E}[U_N(a, b)] \le \mathbb{E}[(f_N - a)^+].$$
+Doob's upcrossing estimate is an important inequality and is central in proving the martingale
+convergence theorems.
+
+## Main definitions
+
+* `measure_theory.upper_crossing a b f N n`: is the stopping time corresponding to `f` crossing
+  above `b` the `n`-th time.
```

Comment:
can you also mention the role of `N`? Also, the name is not so clear as it could be interpreted as a number of upper crossings. What about `upper_crossing_time`? I know it's longer, but since it's not a notion we will use so often, I would favor expressivity over shortness.